Fix deadlock on eventqueue when it's being drained when endpoints are being restored

christarazi opened the pull request:
See commit msgs.

Backporters, please ensure that https://github.com/cilium/cilium/pull/13608 is
backported first, as this PR depends on it.

Fixes: https://github.com/cilium/cilium/pull/13608

This does not require a release-note as this was a regression introduced by
https://github.com/cilium/cilium/pull/13608 which has not made it into any
official release.

christarazi committed:
eventqueue: Add missing godoc

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
eventqueue: Forcefully drain to prevent deadlock

This commit ensures that the EventQueue is fully drained, even when its
not running its loop. When endpoints are being restored, their
EventQueue is initialized, but non-running state (processing events). It
is the job of the endpoint manager to kick off the event loop by calling
Expose() on the endpoint.

This commit fixes the following commits which causes Cilium to be stuck
waiting for the EventQueue to drain (WaitToBeDrained()):

290d9e942 ("daemon: Init endpoint queue during validation")
79bf42515 ("endpoint: Add function to initialize event queue")

Cilium becoming stuck is described in the following flow:
  - Endpoints began restoration
  - Endpoint's EventQueue initialized (but never run)
  - Endpoint's metadata data resolver controller kicked off
  - Visilbity and bandwidth policy events enqueued
  - Endpoint fails restoration due to some issue (e.g. interface not
    found, etc)
  - Endpoint queued for deletion because it failed restoration
  - As part of endpoint deletion, the EventQueue is stopped and drained
  - Cilium deadlocks trying to drain, but the EventQueue run loop was
    never run, which would pop events off the `events` channel, and
    close the `eventsClosed` channel

This commit fixes this deadlock by forcefully running the event loop to
drain the queue. After the `events` channel is closed (from Stop()), the
loop will terminate and the `eventsClosed` channel will close, thereby
unblocking WaitToBeDrained().

Stacktrace from `gops`:

```
goroutine 632 [chan receive, 1 minutes]:
github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).WaitToBeDrained(0xc00013c960)
        /go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:322 +0x1ad
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).Delete(0xc000ad6900, 0x27faee0, 0xc00062ac40, 0x27fba60, 0xc00099a120, 0x2877280, 0xc0005d2340, 0x430101, 0x0, 0x0, ...)
        /go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:2194 +0x91
github.com/cilium/cilium/daemon/cmd.(*Daemon).deleteEndpointQuiet(...)
        /go/src/github.com/cilium/cilium/daemon/cmd/endpoint.go:674
github.com/cilium/cilium/daemon/cmd.(*Daemon).regenerateRestoredEndpoints.func2(0xc00062ac40, 0xc000b63214, 0xc000ad6900)
        /go/src/github.com/cilium/cilium/daemon/cmd/state.go:302 +0x7c
created by github.com/cilium/cilium/daemon/cmd.(*Daemon).regenerateRestoredEndpoints
        /go/src/github.com/cilium/cilium/daemon/cmd/state.go:296 +0x8a0
```

Signed-off-by: Chris Tarazi <chris@isovalent.com>

pchaigno reviewed (approved):
LGTM but would be nice to get some :eyes: more familiar with Go channels.

aanm commented on pkg/eventqueue/eventqueue.go:
```diff
@@ -309,9 +316,19 @@ func (q *EventQueue) WaitToBeDrained() {
+		for ev := range q.events {
+			// Close the results channel on the event to stop receivers from
+			// blocking.
+			close(ev.eventResults)
```
This will have a different behavior, i.e., we are not "canceling" similarly what is being done on line 263.

Discussed offline.

jdef commented on pkg/eventqueue/eventqueue_test.go:
```diff
@@ -213,3 +213,27 @@ func (s *EventQueueSuite) TestEnqueueTwice(c *C) {
+
+	select {
+	case <-res:
+	case <-time.After(5 * time.Second):
```
is 5s standard for timeouts in cilium test cases? seems pretty low for a busy CI system (read: possible test flake generator)

christarazi replied:
I don't think we have a standard timeout in unit tests. Just doing a quick, non-exhaustive grep over the codebase gives:

```shell
❯ rg -i "testtimeout .+ =.+ " --glob '!vendor' -c
pkg/completion/completion_test.go:1
pkg/envoy/xds/server_e2e_test.go:1
proxylib/npds/client_test.go:1
```

It seems like we define a specific timeout (if necessary) for code that is asynchronous, depending on the test itself. In this case though for the eventqueue, the `res` channel should be ready to receive almost immediately because the event is a dummy event. While yes it's possible that on a very contended system, 5 seconds for Go to schedule the `select` might be too little, but the likelihood of that is extremely small. This test has been around for probably over a year now, so it seems to be OK.

christarazi committed:
eventqueue: Sort functions by importance

This commit contains no functional changes. It just reorders some of the
functions by importance in this file so that it's easier to parse. This
should hopefully reduce the number of times needed to scroll up and
down.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
